diff -Naur idl4-1.0.2.pristine/src/arch/v4/ms.cc idl4-1.0.2/src/arch/v4/ms.cc
--- idl4-1.0.2.pristine/src/arch/v4/ms.cc	2003-04-25 20:11:16.000000000 +0200
+++ idl4-1.0.2/src/arch/v4/ms.cc	2007-04-18 09:23:23.000000000 +0200
@@ -907,7 +907,7 @@
   
   return getWrapperReturnType()->buildDeclaration( 
     new CASTDeclarator(wrapperIdentifier, 
-      buildWrapperParams(key)), 
+      buildWrapperParams(key), NULL, NULL, NULL, buildWrapperAttributes()), 
     compound
   );
 }
diff -Naur idl4-1.0.2.pristine/src/cast/cast.cc idl4-1.0.2/src/cast/cast.cc
--- idl4-1.0.2.pristine/src/cast/cast.cc	2003-08-25 06:41:22.000000000 +0200
+++ idl4-1.0.2/src/cast/cast.cc	2007-04-18 09:23:23.000000000 +0200
@@ -361,13 +361,22 @@
 
   if (qualifiers)
     {
+      print(" ");
       qualifiers->writeAll(" ");
-      if (identifier)
-        print(" ");
     }  
 
+  if (attributes)
+    {
+      print(" ");
+      attributes->writeAll(" ");
+    }
+    
+
   if (identifier) 
-    identifier->write();
+    {
+      print(" ");
+      identifier->write();
+    }
 
   if (subdecl)
     {
@@ -407,12 +416,6 @@
       bitSize->write();
     }
 
-  if (attributes)
-    {
-      print(" ");
-      attributes->writeAll(" ");
-    }
-    
   if (initializer)
     {
       if (identifier)
@@ -427,15 +430,15 @@
 void CASTIntegerConstant::write()
 
 {
-  enter("integer constant: %Ld", value);
+  enter("integer constant: %lld", value);
 
   if (isUnsigned)
-    print("%Ldu", value);
-    else print("%d", value);
+    print("%llu", value);
+    else print("%lld", value);
   if (isLong)
     print("l");
   
-  leave("integer constant: %Ld", value);
+  leave("integer constant: %lld", value);
 }
 
 void CASTStringConstant::write()
@@ -804,18 +807,18 @@
 void CASTHexConstant::write()
 
 {
-  enter("hex constant: 0x%LX", value);
+  enter("hex constant: 0x%llX", value);
 
   if (value>1)
-    print("0x%LX", value);
-    else print("%d", value);
+    print("0x%llX", value);
+    else print("%lld", value);
 
   if (isUnsigned)
     print("u");
   if (isLong)
     print("l");
     
-  leave("hex constant: 0x%LX", value);
+  leave("hex constant: 0x%llX", value);
 }
 
 void CASTAsmStatement::write()
diff -Naur idl4-1.0.2.pristine/src/include/user/idl4/api/v4/interface.h idl4-1.0.2/src/include/user/idl4/api/v4/interface.h
--- idl4-1.0.2.pristine/src/include/user/idl4/api/v4/interface.h	2003-08-25 06:44:55.000000000 +0200
+++ idl4-1.0.2/src/include/user/idl4/api/v4/interface.h	2007-04-18 09:32:17.000000000 +0200
@@ -30,7 +30,12 @@
   L4_Acceptor_t _rcv_window;
 } CORBA_Environment;
 
-#define idl4_default_environment ((CORBA_Environment){ CORBA_NO_EXCEPTION, 0, NULL, L4_Timeouts(L4_Never, L4_Never), L4_UntypedWordsAcceptor })
+#define idl4_default_environment ((CORBA_Environment) {			\
+				  _major: CORBA_NO_EXCEPTION,           \
+					  _minor: 0,			\
+					  _data: NULL,			\
+					  _timeout: L4_Timeouts(L4_Never, L4_Never), \
+					  _rcv_window: { raw: 0 } })
 
 static inline void CORBA_exception_set(idl4_server_environment *ev, CORBA_exception_type minor, void *param)
 
